### PR TITLE
Rework protokube dns so it shares code with dns-controller

### PR DESCRIPTION
**Pull request by justinsb**
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/697)

### CODE REVIEW

File: protokube/cmd/protokube/main.go [hunk 1]
```diff
@@ -20,22 +20,48 @@ import (
...
+	"github.com/spf13/pflag"
+	"k8s.io/kops/dns-controller/pkg/dns"
 	"k8s.io/kops/protokube/pkg/protokube"
+	"k8s.io/kubernetes/federation/pkg/dnsprovider"
 	"net"
 	"os"
 	"strings"
 )
 
+var (
```
Comment (JuJu227) [inline protokube/cmd/protokube/main.go]: You could just put them in the same block 
`var(
          flags = .....
          Buildin......
     )   
`

File: protokube/cmd/protokube/main.go [hunk 2]
```diff
@@ -75,10 +105,10 @@ func main() {
 		dnsInternalSuffix = "." + dnsInternalSuffix
 	}
 
-	if dnsZoneName == "" {
-		tokens := strings.Split(dnsInternalSuffix, ".")
-		dnsZoneName = strings.Join(tokens[len(tokens)-2:], ".")
-	}
+	//if dnsZoneName == "" {
```
Comment (JuJu227) [inline protokube/cmd/protokube/main.go]: Let's not forget to delete dead code

**Comment by qqshfox** [comment]
I'm trying to implement a new dns provider to replace route53 in China. Is this PR getting merged anytime soon? I think this can make the new implementation much simpler.

**Comment by kris-nova** [comment]
We would need a rebase first - @justinsb can you shed some light here? I will take over ownership of the merge, but need to know what your thoughts on the matter are.. 

Please bear with us @qqshfox

**Comment by chrislovecnm** [comment]
@k8s-bot aws e2e test this

**Comment by chrislovecnm** [comment]
and e2e is failing ...

**Comment by justinsb** [comment]
@qqshfox I am planning on getting this in :-)  If you want to contribute a dnsprovider into kubernetes that works in china - that would be great!  It would go here: https://github.com/kubernetes/kubernetes/tree/master/federation/pkg/dnsprovider/providers

How are you thinking of implementing it?  And does AWS China allow _private_ route53 hosted zones?

**Comment by justinsb** [comment]
So now we are packaging our own version of protokube, I think this is finally unblocked

File: protokube/cmd/protokube/main.go [hunk 3]
```diff
@@ -45,20 +73,24 @@ func main() {
 	flagChannels := ""
 	flag.StringVar(&flagChannels, "channels", flagChannels, "channels to install")
 
+	// Trick to avoid 'logging before flag.Parse' warning
+	flag.CommandLine.Parse([]string{})
+
 	flag.Set("logtostderr", "true")
-	flag.Parse()
+
+	flags.AddGoFlagSet(flag.CommandLine)
```
Comment (chrislovecnm) [inline protokube/cmd/protokube/main.go]: we should have an issue for using cobra here

File: protokube/cmd/protokube/main.go [hunk 4]
```diff
@@ -89,10 +116,34 @@ func main() {
...
+		}
+
+		zoneRules, err := dns.ParseZoneRules(zones)
+		if err != nil {
+			return fmt.Errorf("unexpected zone flags: %q", err)
+		}
+
+		dnsController, err = dns.NewDNSController(dnsProvider, zoneRules)
+		if err != nil {
+			return err
```
Comment (chrislovecnm) [inline protokube/cmd/protokube/main.go]: should we `fmt` the error here?

Comment (justinsb) [inline protokube/cmd/protokube/main.go]: What do you suggest?

File: protokube/cmd/protokube/main.go (hunk 3 above)
Comment (justinsb) [inline protokube/cmd/protokube/main.go]: Not sure why?

File: dns-controller/pkg/dns/dnscontroller.go [hunk 5]
```diff
@@ -32,6 +32,8 @@ import (
 	"k8s.io/kubernetes/federation/pkg/dnsprovider/rrstype"
 )
 
+const DefaultTTL = time.Minute
```
Comment (kris-nova) [inline dns-controller/pkg/dns/dnscontroller.go]: One of these days we are going to have a fabulous configuration something or other for this stuff - minute sounds good though.. but I know plenty of DNS fans who would have something to say about this :)

File: upup/pkg/fi/cloudup/template_functions.go [hunk 6]
```diff
@@ -211,7 +211,9 @@ func (tf *TemplateFunctions) DnsControllerArgv() ([]string, error) {
 	}
 	// permit wildcard updates
 	argv = append(argv, "--zone=*/*")
-	argv = append(argv, "-v=8")
+
+	// Verbose, but not crazy logging
+	argv = append(argv, "-v=2")
```
Comment (kris-nova) [inline upup/pkg/fi/cloudup/template_functions.go]: Suggest 4 and we can bump it down later - not a deal breaker for merge

**Comment by kris-nova** [comment]
@k8s-bot aws e2e test this

File: upup/pkg/fi/cloudup/template_functions.go (hunk 6 above)
Comment (justinsb) [inline upup/pkg/fi/cloudup/template_functions.go]: It's so easy to change though if someone wants more logging.  Also I don't want to wait for the tests to rerun.  I'll push this with v=2 if that's OK, and then we can revise if needed.

**Comment by qqshfox** [comment]
@justinsb Sorry for late reply. Thank you for have this PR merged. I've made a little progress with some heavily handcrafted and hardcoded modifications. I implemented a new dnsprovider named dnspod [(Chinese website)](https://www.dnspod.cn) which claimed it's the large domain service provider in China.  I've managed to launch a working cluster after dealing with a few AWS China Region specific problems, e.g. internet access issue (never succeeded to download the k8s artifacts, docker deb, etc.), no NatGateway, no Route53 (even private zones).

Since I'm new to the kops and k8s codebases, I'm struggling with some odd behaviors that might be related with the DNSZone and Private Topology now. I hope I can submit a WIP PR in the next a few days after clean up the mess.

**Comment by justinsb** [comment]
@qqshfox that would be great - do feel free to submit a WIP PR that is as messy as you need and I'll help - I would love to solve the DNS-in-china problem.

Also if you have any notes on the artifacts and how you solved them, it would be great if you opened an issue and described what you did.  Even if they are just sketch-notes, that would be very useful :-)

**Pull request merged**